libtiledbsoma/src/soma/soma_array.cc

```diff
@@ -1205,13 +1205,32 @@ uint64_t SOMAArray::nnz() {
     // compute total_cell_num while going through the loop
     uint64_t total_cell_num = 0;
     std::vector<std::array<uint64_t, 2>> non_empty_domains(fragment_count);
+
+    // The loop after this only works if dim 0 is int64 soma_joinid.
+    // That's the case for _almost_ all SOMADataFrame objects, but
+    // not the "variant-indexed" ones: the SOMA spec only requires
+    // that soma_joinid be present as a dim or an attr.
+    auto dim = tiledb_schema()->domain().dimension(0);
+    auto dim_name = dim.name();
+    auto type_code = dim.type();
+    if (dim_name != "soma_joinid" || type_code != TILEDB_INT64) {
```

Comment:
Shouldn't the "fast" version work for any int64 dimension whether or not it's the `soma_joinid`?

Reply:
if they have a dim which is of type `int64` but it isn't `soma_joinid` -- which is custom -- then we should be doing cell-counting i think

Reply:
In theory this check is redundant then since `soma_joinid` is always `int64`, but it doesn't hurt to have the extra check.